MobileCenter/MobileCenter/Internals/Device/MSDeviceTracker.m

```diff
@@ -14,43 +16,103 @@
 } ms_info_t;
 
 // SDK versioning.
-ms_info_t mobilecenter_library_info __attribute__((section("__TEXT,__ms_ios,regular,no_dead_strip"))) = {
-    .info_version = 1,
-    .ms_name = MOBILE_CENTER_C_NAME,
-    .ms_version = MOBILE_CENTER_C_VERSION,
-    .ms_build = MOBILE_CENTER_C_BUILD
-};
+ms_info_t mobilecenter_library_info
+    __attribute__((section("__TEXT,__ms_ios,regular,no_dead_strip"))) = {.info_version = 1,
+                                                                         .ms_name = MOBILE_CENTER_C_NAME,
+                                                                         .ms_version = MOBILE_CENTER_C_VERSION,
+                                                                         .ms_build = MOBILE_CENTER_C_BUILD};
 
-@interface MSDeviceTracker()
+static NSString *const kMSPastDevicesKey = @"pastDevicesKey";
+static NSUInteger const kMSMaxDevicesHistoryCount = 5;
 
-// We need a private setter for the device to avoid ivars direct access warning.
+@interface MSDeviceTracker ()
+
+// We need a private setter for the device to avoid the warning that is related to direct access of ivars.
 @property(nonatomic) MSDevice *device;
 
 @end
 
 @implementation MSDeviceTracker : NSObject
 
-@synthesize device = _device;
-
-static MSWrapperSdk *wrapperSdkInformation = nil;
 static BOOL needRefresh = YES;
+static MSWrapperSdk *wrapperSdkInformation = nil;
+
+- (instancetype)init {
+  if (self = [super init]) {
+
+    // Restore past sessions from NSUserDefaults.
+    NSData *devices = [MS_USER_DEFAULTS objectForKey:kMSPastDevicesKey];
+    if (devices != nil) {
+      NSArray *arrayFromData = [NSKeyedUnarchiver unarchiveObjectWithData:devices];
+
+      // If array is not nil, create a mutable version.
+      if (arrayFromData)
+        _deviceHistory = [NSMutableArray arrayWithArray:arrayFromData];
+    }
+
+    // Create new array and creade device info in case we don't have any from disk.
+    if (_deviceHistory == nil) {
+      _deviceHistory = [NSMutableArray<MSDeviceHistoryInfo *> new];
+
+      // Don't assign the new device to the property to continue using lazy initialization later.
+      // We're creating this one to have a history.
+      [self device];
```

Comment:
I just noticed but Isn't this and line 62 breaking the lazy init pattern? Because as the _device is nil for now the [self device] will init it.